### PR TITLE
Initial implementation of variable difficulty

### PR DESCRIPTION
**Pull request by mugling**
The current slow/fast/tough zombie mods are a bad design:
* Only support arbitrary changes (typically 50% and 200%)
* Currently dependent on `dda` content pack
* Needlessly tested in `--check-mods`
* Only available for Lua builds

This PR exposes them as percentages on worldgen.
The next step is to support difficulty presets, *'Easy'*, *'Medium'*, *'Hard'* etc.

difficulty

### CODE REVIEW

File: src/options.h [hunk 1]
```diff
@@ -172,7 +174,8 @@ class options_manager : public JsonSerializer, public JsonDeserializer
         void add( const std::string sNameIn, const std::string sPageIn,
                   const std::string sMenuTextIn, const std::string sTooltipIn,
                   const int iMinIn, int iMaxIn, int iDefaultIn,
-                  copt_hide_t opt_hide = COPT_NO_HIDE );
+                  copt_hide_t opt_hide = COPT_NO_HIDE,
+                  const std::string &format = "%i" );
```
Comment (codemime) [inline src/options.h]: `options_manager` clearly needs refactoring: 9th parameter was added to two of six monstrous overloaded functions.

Comment (mugling) [inline src/options.h]: Could not possibly agree more. It's a horrendously bad design and should probably be replaced en-masse. I'm looking into this for a separate PR.

For now though a minimal number of lines gets rid of a worse evil (spurious mods) and as a bonus provides two meaningful configuration options.

**Comment by harald921** [comment]
>Allow free scaling of monster hp at worldgen

Mugling! <3

**Comment by Coolthulhu** [comment]
Monster hp should never reach 0. Rounding would be preferable to truncation there.

File: src/options.cpp [hunk 2]
```diff
@@ -410,17 +414,10 @@ std::string options_manager::cOpt::getValue() const
...
+        return string_format( format, iSet );
 
     } else if (sType == "float") {
-        std::stringstream ssTemp;
-        ssTemp.imbue(std::locale::classic());
-        const int precision = (fStep >= 0.09) ? 1 : (fStep >= 0.009) ? 2 : (fStep >= 0.0009) ? 3 : 4;
-        ssTemp.precision(precision);
-        ssTemp << std::fixed << fSet;
-        return ssTemp.str();
+        return string_format( format, fSet );
```
Comment (Coolthulhu) [inline src/options.cpp]: Are you sure it won't affect precision?

Comment (mugling) [inline src/options.cpp]: We don't have any options with more than two decimals precision

**Comment by mugling** [comment]
> Monster hp should never reach 0. Rounding would be preferable to truncation there.

Good point, needs a lower bound

**Comment by mugling** [comment]
> Monster hp should never reach 0. Rounding would be preferable to truncation there.
> 
> Good point, needs a lower bound

Fixed

**Comment by Coolthulhu** [comment]
Annoyance: the query keeps the '%' sign from the display. This means you can't just add '0' at the end of it, because "100%0" is just "100" and not "1000".

**Comment by mugling** [comment]
> Annoyance: the query keeps the '%' sign from the display. This means you can't just add '0' at the end of it, because "100%0" is just "100" and not "1000".

Can you rephrase?

**Comment by Coolthulhu** [comment]
* Go to world options
* Select monster speed
* Press enter to edit it

Expected:
* Query displays 100, adding 0 at the end and pressing enter sets it to 1000%

Got:
* Query displays 100%, that % sign is useless and only makes parsing stop at it

**Comment by mugling** [comment]
I didn't know you could press `Enter`!

**Comment by Coolthulhu** [comment]
It doesn't work in main menu for some reason. But in a loaded world it does and it much better than holding left/right.

**Comment by harald921** [comment]
Is it possible to update and change the variable HP without regenerating the world? Mostly curious whether or not I will be able to tweak the settings without regenerating all the time.

**Comment by Coolthulhu** [comment]
Yes, I did it during testing.
You need to save and reload the world, though.

**Comment by mugling** [comment]
> I didn't know you could press Enter!

Where does the code for this live. I'm looking in `world_factory.cpp` but it's not obvious

**Comment by mugling** [comment]
Found a candidate in `options.cpp`

**Comment by mugling** [comment]
> Go to world options
>     Select monster speed
>     Press enter to edit it
> 
> Expected:
> 
>     Query displays 100, adding 0 at the end and pressing enter sets it to 1000%
> 
> Got:
> 
>     Query displays 100%, that % sign is useless and only makes parsing stop at it

Nothing happens when I press `ENTER`

**Comment by sick-trigger** [comment]
@mugling could this be changed to only affect zombie monsters? I liked the balance of fast+tough zombies but doing the same for all monsters is a little overwhelming

**Comment by mugling** [comment]
> @mugling could this be changed to only affect zombie monsters? I liked the balance of fast+tough zombies but doing the same for all monsters is a little overwhelming

Not easily
